### PR TITLE
CYS - Core:  Add async font installation

### LINKED ISSUE #44257: [CYS on Core] Async Fonts installation whenever the user grants consent
As part of our work with the CYS, we have to ensure all custom fonts are properly installed and available in the Assembler by the time the store owner reaches the customization phase.

In order to ensure optimal performance, we need to detach the download of fonts from the CYS loading screen, relying on a scheduled action for installing them whenever the user accepts the terms & conditions (whenever WooCommerce is installed & the permission is granted), combined with an emergency fallback fonts installation during the CYS loading screen to be triggered only if something went wrong with the initial async installation.

By following this approach, we ensure persistent performance optimization, as the loading time has the potential to grow even further if more fonts are downloaded and added to the mix.

## The Scope

- Make sure a new action is scheduled whenever the [`woocommerce_allow_tracking`](https://github.com/woocommerce/woocommerce/blob/4df15fe948d8a5f85b68879e57652772bf013425/plugins/woocommerce-admin/client/core-profiler/index.tsx#L209) option is saved and/or updated in the database, and set to `true` (`yes`). Make sure the action is scheduled to run immediately.
- Attach the method responsible for the fonts installation to this scheduled action. Considering Gutenberg and the new Font Library still doesn't have an API on the PHP side, at this first moment we are porting over any code that might be needed while simultaneously kicking off a conversation with folks on GB and identifying the possibility of an upstream contribution.
- The following code is an example for scheduling the action:

https://github.com/woocommerce/woocommerce/blob/a38c7bcccfab789e7555e9b7c34dc7d15fdd4228/plugins/woocommerce/src/Blocks/BlockPatterns.php#L317-L331

- During the CYS loading screen, add validation to ensure all relevant fonts are in place and were successfully installed. - If for some reason these fonts are not available, trigger the fonts installation on the JS side so the store owner have them all available by the time they reach the Assembler.

### PR DESCRIPTION
**Pull request by gigitux**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This PR introduces an async job to install the fonts. 

The job starts when the `woocommerce_allow_tracking'` is set to yes and the Font Library is available (https://github.com/woocommerce/woocommerce/pull/44473/files#diff-9ff9cbf9fc5ce9eb0df48369788b45884e619f1841b94e4e3fce7bcb75c14e9fR104-R113).

The async job:
- gets the font info from the google font collection.
- create `wp_font_family` posts.
- download the various font_faces, upload them to the WordPress site, and create the `wp_font_face` posts.

This PR introduces two internal classes:
- `FontFace`
- `FontFamily` 

Both are utils classes to create and install the font families and font faces. As it is annotated in the code, we have to remove them and refactor this code when the WordPress PHP API to install fonts is available: https://github.com/WordPress/gutenberg/issues/58670

I think that we should add unit tests for the code that we added in this PR and in https://github.com/woocommerce/woocommerce/pull/44004 to ensure that we don't introduce any regression in the logic to install the fonts. What do you think?

Closes #44257 .

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

#### These testing instructions require a connection with the database

1. Create a new WooCommerce installation with this version.
2. Ensure the WooCommerce Beta Tester plugin is installed and activated (available on this monorepo).
3. Head over to `/wp-admin/tools.php?page=woocommerce-admin-test-helper` and enable `customize-store` feature flag.
4. Install this build of Gutenberg that contains the new Font Library API: [gutenberg.zip](https://github.com/woocommerce/woocommerce/files/14236920/gutenberg.zip)
5. Connect to the database and ensure that doesn't exist any post with these post-type: `wp_font_family`, `wp_font_face` and `wp_global_styles`.
6. Go on `/wp-admin/admin.php?page=wc-settings&tab=advanced&section=woocommerce_com` and enable the tracking.
7. Go on `/wp-admin/admin.php?page=wc-status&tab=action-scheduler` and ensure that `woocommerce_install_assembler_fonts` is in pending.
8. Ensure that the `woocommerce_install_assembler_fonts` job doesn't fail.
9. Visit the `wp-admin/admin.php?page=wc-admin&path=/customize-store`.
10. Open the network tools.
11. Follow the process.
12. Ensure that no POST requests are done on the `/font-faces/` endpoint. (it is expected that GET requests are done).
13. Wait until the Assembler is loaded.
14. Due [to this bug](https://github.com/WordPress/gutenberg/issues/58765), it is necessary to refresh the page to ensure that the fonts are loaded.
15. Click on "Change Fonts".
16. Ensure that the right fonts are applied.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/src/Internal/Admin/Onboarding/OnboardingFonts.php [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+		);
+	}
+
+		/**
+		 * Get font family by slug from font collection.
+		 *
+		 * @param string $slug Font slug.
+		 * @param array  $font_families_collection Font families collection.
+		 * @return array|null
+		 */
```
Comment (albarin) [inline plugins/woocommerce/src/Internal/Admin/Onboarding/OnboardingFonts.php]: Nit: the comment is misaligned.

**Review by albarin — APPROVED** [review]
Great work 🙇

**Review by nefeline — COMMENTED** [review]
Thanks for your great work here @gigitux !

I can confirm the async fonts installation is working as expected when following the testing instructions on this PR, well done! 👏 

However, it looks like we are still not preventing the fonts installation whenever the user did not opt in to tracking on the CYS loading screen: only the scheduled action has a check for the value of `'woocommerce_allow_tracking'`. Could we make a change here to prevent this from happening?

**Comment by gigitux** [comment]
> Thanks for your great work here @gigitux !
> 
> I can confirm the async fonts installation is working as expected when following the testing instructions on this PR, well done! 👏
> 
> However, it looks like we are still not preventing the fonts installation whenever the user did not opt in to tracking on the CYS loading screen: only the scheduled action has a check for the value of `'woocommerce_allow_tracking'`. Could we make a change here to prevent this from happening?

Thanks for the review!

Good point! Given that this PR is already significant, I created a dedicated issue: https://github.com/woocommerce/woocommerce/issues/44552. I'm going to work on it!

**Review by nefeline — APPROVED** [review]
> Good point! Given that this PR is already significant, I created a dedicated issue: https://github.com/woocommerce/woocommerce/issues/44552. I'm going to work on it!

Alright, thanks @gigitux ! I'm going ahead and approving the PR given the fix will be worked on separately. Thanks again for your work here 🙌

**Pull request merged**